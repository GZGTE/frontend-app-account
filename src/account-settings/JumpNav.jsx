import React from 'react';
import { injectIntl, intlShape } from '@edx/frontend-i18n'; // eslint-disable-line
import { NavHashLink } from 'react-router-hash-link';
import Scrollspy from 'react-scrollspy';

import messages from './AccountSettingsPage.messages';


function JumpNav({ intl }) {
  return (
    <div className="jump-nav">
      <Scrollspy
        items={[
          'basic-information',
          'profile-information',
          'social-media',
          'site-preferences',
          'linked-accounts',
          'delete-account',
        ]}
        className="list-unstyled"
        currentClassName="font-weight-bold"
      >
        <li>
          <NavHashLink to="#basic-information">
            {intl.formatMessage(messages['account.settings.section.account.information'])}
          </NavHashLink>
        </li>
        <li>
          <NavHashLink to="#profile-information">
            {intl.formatMessage(messages['account.settings.section.profile.information'])}
          </NavHashLink>
        </li>
        <li>
          <NavHashLink to="#social-media">
            {intl.formatMessage(messages['account.settings.section.social.media'])}
          </NavHashLink>
        </li>
        <li>
          <NavHashLink to="#site-preferences">
            {intl.formatMessage(messages['account.settings.section.site.preferences'])}
          </NavHashLink>
        </li>
        <li>
          <NavHashLink to="#linked-accounts">
            {intl.formatMessage(messages['account.settings.section.linked.accounts'])}
          </NavHashLink>
        </li>
        <li>
          <NavHashLink to="#delete-account">
            {intl.formatMessage(messages['account.settings.delete.account.header'])}
          </NavHashLink>
        </li>
      </Scrollspy>
    </div>
  );
}


JumpNav.propTypes = {
  intl: intlShape.isRequired,
};


export default injectIntl(JumpNav);
