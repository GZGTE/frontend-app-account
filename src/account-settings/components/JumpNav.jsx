import React from 'react';
import { injectIntl, intlShape } from '@edx/frontend-i18n'; // eslint-disable-line
import { NavHashLink } from 'react-router-hash-link';

import messages from '../AccountSettingsPage.messages';


function JumpNav({ intl }) {
  return (
    <div className="jump-nav">
      <ul className="list-unstyled">
        <li>
          <NavHashLink to="#basic-information">
            {intl.formatMessage(messages['account.settings.section.account.information'])}
          </NavHashLink>
        </li>
        <li>
          <NavHashLink to="#profile-information">
            {intl.formatMessage(messages['account.settings.section.profile.information'])}
          </NavHashLink>
        </li>
        <li>
          <NavHashLink to="#social-media">
            {intl.formatMessage(messages['account.settings.section.social.media'])}
          </NavHashLink>
        </li>
        <li>
          <NavHashLink to="#site-preferences">
            {intl.formatMessage(messages['account.settings.section.site.preferences'])}
          </NavHashLink>
        </li>
        <li>
          <NavHashLink to="#linked-accounts">
            {intl.formatMessage(messages['account.settings.section.linked.accounts'])}
          </NavHashLink>
        </li>
      </ul>
    </div>
  );
}


JumpNav.propTypes = {
  intl: intlShape.isRequired,
};


export default injectIntl(JumpNav);
