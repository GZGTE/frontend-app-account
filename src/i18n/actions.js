export const SET_LOCALE = 'I18N__SET_LOCALE';

export const setLocale = locale => ({
  type: SET_LOCALE,
  payload: {
    locale,
  },
});
