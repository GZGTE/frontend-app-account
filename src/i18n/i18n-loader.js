/**
 * For each locale we want to support, react-intl needs 1) the locale-data, which includes
 * information about how to format numbers, handle plurals, etc., and 2) the translations, as an
 * object holding message id / translated string pairs.  A locale string and the messages object are
 * passed into the IntlProvider element that wraps your element hierarchy.
 *
 * Note that react-intl has no way of checking if the translations you give it actually have
 * anything to do with the locale you pass it; it will happily use whatever messages object you pass
 * in.  However, if the locale data for the locale you passed into the IntlProvider was not
 * correctly installed with addLocaleData, all of your translations will fall back to the default
 * (in our case English), *even if you gave IntlProvider the correct messages object for that
 * locale*.
 */

import { addLocaleData } from 'react-intl';
import Cookies from 'universal-cookie';

import arLocale from 'react-intl/locale-data/ar';
import enLocale from 'react-intl/locale-data/en';
import es419Locale from 'react-intl/locale-data/es';
import frLocale from 'react-intl/locale-data/fr';
import zhcnLocale from 'react-intl/locale-data/zh';
import caLocale from 'react-intl/locale-data/ca';
import heLocale from 'react-intl/locale-data/he';
import idLocale from 'react-intl/locale-data/id';
import kokrLocale from 'react-intl/locale-data/ko';
import plLocale from 'react-intl/locale-data/pl';
import ptbrLocale from 'react-intl/locale-data/pt';
import ruLocale from 'react-intl/locale-data/ru';
import thLocale from 'react-intl/locale-data/th';
import ukLocale from 'react-intl/locale-data/uk';

import COUNTRIES, { langs as countryLangs } from 'i18n-iso-countries';
import LANGUAGES, { langs as languageLangs } from '@cospired/i18n-iso-languages';

import arMessages from './messages/ar.json';
import caMessages from './messages/ca.json';
// no need to import en messages-- they are in the defaultMessage field
import es419Messages from './messages/es_419.json';
import frMessages from './messages/fr.json';
import zhcnMessages from './messages/zh_CN.json';
import heMessages from './messages/he.json';
import idMessages from './messages/id.json';
import kokrMessages from './messages/ko_kr.json';
import plMessages from './messages/pl.json';
import ptbrMessages from './messages/pt_br.json';
import ruMessages from './messages/ru.json';
import thMessages from './messages/th.json';
import ukMessages from './messages/uk.json';

addLocaleData([
  ...arLocale,
  ...enLocale,
  ...es419Locale,
  ...frLocale,
  ...zhcnLocale,
  ...caLocale,
  ...heLocale,
  ...idLocale,
  ...kokrLocale,
  ...plLocale,
  ...ptbrLocale,
  ...ruLocale,
  ...thLocale,
  ...ukLocale,
]);

// TODO: When we start dynamically loading translations only for the current locale, change this.
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/ar.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/en.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/es.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/fr.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/zh.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/ca.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/he.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/id.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/ko.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/pl.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/pt.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/ru.json'));
// COUNTRIES.registerLocale(require('i18n-iso-countries/langs/th.json')); // Doesn't exist in lib.
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/uk.json'));

// TODO: When we start dynamically loading translations only for the current locale, change this.
// TODO: Also note that a bunch of languages are missing here. They're present but commented out
// for reference. That's because they're not implemented in this library.  If you read this and it's
// been a while, go check and see if that's changed!
// LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/ar.json'));
LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/en.json'));
LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/es.json'));
LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/fr.json'));
// LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/zh.json'));
// LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/ca.json'));
// LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/he.json'));
// LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/id.json'));
// LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/ko.json'));
LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/pl.json'));
LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/pt.json'));
// LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/ru.json'));
// LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/th.json'));
// LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/uk.json'));

const messages = {
  // current fallback strategy is to use the first two letters of the locale code
  ar: arMessages,
  es: es419Messages,
  fr: frMessages,
  zh: zhcnMessages,
  ca: caMessages,
  he: heMessages,
  id: idMessages,
  ko: kokrMessages,
  pl: plMessages,
  pt: ptbrMessages,
  ru: ruMessages,
  th: thMessages,
  uk: ukMessages,
};

const serverCodeLookup = {
  ar: 'ar',
  en: 'en',
  es: 'es-419',
  fr: 'fr',
  zh: 'zh-cn',
  ca: 'ca',
  he: 'he',
  id: 'id',
  ko: 'ko-kr',
  pl: 'pl',
  pt: 'pt-br',
  ru: 'ru',
  th: 'th',
  uk: 'uk',
};

/**
 * TODO: This function is a hack.  Because the client is using two-letter codes and the server
 * is using fully qualified codes with a locale, if we want the server to understand what code
 * we're using, we need to convert back to whatever it was originally using.  This will
 * immediately break down if we have another version of Spanish (es), Chinese (zh), or
 * Portugese (pt), for instance. That's why this is "assumed".
 */
const getAssumedServerLanguageCode = lang => (serverCodeLookup[lang] !== undefined ?
  serverCodeLookup[lang] : lang);

const cookies = new Cookies();

const getTwoLetterLanguageCode = code => code.substr(0, 2);

// Get the locale by setting priority. Skip if we don't support that language.
const getLocale = (localeStr) => {
  // 1. Explicit application request
  if (localeStr && messages[localeStr] !== undefined) {
    return localeStr;
  }
  // 2. User setting in cookie
  const cookieLangPref = cookies.get(process.env.LANGUAGE_PREFERENCE_COOKIE_NAME);
  if (cookieLangPref && messages[getTwoLetterLanguageCode(cookieLangPref)] !== undefined) {
    return getTwoLetterLanguageCode(cookieLangPref);
  }
  // 3. Browser language (default)
  return getTwoLetterLanguageCode(window.navigator.language);
};

const getMessages = (locale = getLocale()) => messages[locale];

const rtlLocales = ['ar', 'he', 'fa', 'ur'];
const isRtl = locale => rtlLocales.includes(locale);

const handleRtl = () => {
  if (process.env.ENVIRONMENT === 'production') {
    if (isRtl(getLocale())) {
      document.getElementsByTagName('html')[0].setAttribute('dir', 'rtl');
      document.styleSheets[0].disabled = true;
      document.styleSheets[1].disabled = false;
    } else {
      document.getElementsByTagName('html')[0].removeAttribute('dir');
      document.styleSheets[0].disabled = false;
      document.styleSheets[1].disabled = true;
    }
  }
};

/**
 * Provides a lookup table of country IDs to country names for the current locale.
 */
const getCountryMessages = (locale) => {
  const twoLetterLocale = getTwoLetterLanguageCode(locale);
  const finalLocale = countryLangs().includes(twoLetterLocale) ? twoLetterLocale : 'en';

  return COUNTRIES.getNames(finalLocale);
};

/**
 * Provides a lookup table of language IDs to language names for the current locale.
 */
const getLanguageMessages = (locale) => {
  const twoLetterLocale = getTwoLetterLanguageCode(locale);
  const finalLocale = languageLangs().includes(twoLetterLocale) ? twoLetterLocale : 'en';

  return LANGUAGES.getNames(finalLocale);
};

const sortFunction = (a, b) => {
  // If localeCompare exists, use that.  (Not supported in some older browsers)
  if (typeof String.prototype.localeCompare === 'function') {
    return a[1].localeCompare(b[1], getLocale());
  }
  if (a[1] === b[1]) {
    return 0;
  }
  // Otherwise make a best effort.
  return a[1] > b[1] ? 1 : -1;
};

/**
 * Provides a list of countries represented as objects of the following shape:
 *
 * {
 *   key, // The ID of the country
 *   name // The localized name of the country
 * }
 *
 * The list is sorted alphabetically in the current locale.
 * This is useful for select dropdowns primarily.
 */
const getCountryList = (locale) => {
  const countryMessages = getCountryMessages(locale);
  return Object.entries(countryMessages)
    .sort(sortFunction)
    .map(([code, name]) => ({ code, name }));
};

/**
 * Provides a list of languages represented as objects of the following shape:
 *
 * {
 *   key, // The ID of the language
 *   name // The localized name of the language
 * }
 *
 * The list is sorted alphabetically in the current locale.
 * This is useful for select dropdowns primarily.
 */
const getLanguageList = (locale) => {
  const languageMessages = getLanguageMessages(locale);
  return Object.entries(languageMessages)
    .sort(sortFunction)
    .map(([code, name]) => ({ code, name }));
};

export {
  getCountryList,
  getCountryMessages,
  getLanguageList,
  getLanguageMessages,
  getLocale,
  getAssumedServerLanguageCode,
  getMessages,
  handleRtl,
  isRtl,
};
