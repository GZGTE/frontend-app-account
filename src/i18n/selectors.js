// eslint-disable-next-line import/prefer-default-export
export const localeSelector = state => state.i18n.locale;
