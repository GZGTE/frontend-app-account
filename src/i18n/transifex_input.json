{
  "account.settings.page.heading": "Account Settings",
  "account.settings.loading.message": "Loading",
  "account.settings.loading.error": "Error: {error}",
  "account.settings.section.account.information": "Account Information",
  "account.settings.section.account.information.description": "These settings include basic information about your account.",
  "account.settings.field.username": "Username",
  "account.settings.field.full.name": "Full name",
  "account.settings.field.email": "Email address (Sign in)",
  "account.settings.field.email.confirmation": "Weâ€™ve sent a confirmation message to {value}. Click the link in the message to update your email address.",
  "account.settings.email.field.confirmation.header": "Pending confirmation",
  "account.settings.field.dob": "Year of birth",
  "account.settings.field.country": "Country",
  "account.settings.field.site.language": "Site Language",
  "account.settings.field.site.language.help.text": "The language used throughout this site. This site is currently available in a limited number of languages.",
  "account.settings.field.education": "Education",
  "account.settings.field.education.levels.null": "Select a level of education",
  "account.settings.field.education.levels.p": "Doctorate",
  "account.settings.field.education.levels.m": "Master's or professional degree",
  "account.settings.field.education.levels.b": "Bachelor's Degree",
  "account.settings.field.education.levels.a": "Associate's degree",
  "account.settings.field.education.levels.hs": "Secondary/high school",
  "account.settings.field.education.levels.jhs": "Junior secondary/junior high/middle school",
  "account.settings.field.education.levels.el": "Elementary/primary school",
  "account.settings.field.education.levels.none": "No formal education",
  "account.settings.field.education.levels.o": "Other education",
  "account.settings.field.gender": "Gender",
  "account.settings.field.gender.options.null": "Select a gender",
  "account.settings.field.gender.options.f": "Female",
  "account.settings.field.gender.options.m": "Male",
  "account.settings.field.gender.options.o": "Other",
  "account.settings.field.language.proficiencies": "Spoken Languages",
  "account.settings.section.social.media": "Social Media Links",
  "account.settings.section.social.media.description": "Optionally, link your personal accounts to the social media icons on your edX profile.",
  "account.settings.field.social.platform.name.linkedin": "LinkedIn",
  "account.settings.field.social.platform.name.twitter": "Twitter",
  "account.settings.field.social.platform.name.facebook": "Facebook",
  "account.settings.editable.field.password.reset.button": "We've sent a message to {email}. Click the link in the message to reset your password. Didn't receive the message? Contact {technicalSupportLink}.",
  "account.settings.editable.field.action.save": "Save",
  "account.settings.editable.field.action.cancel": "Cancel",
  "account.settings.editable.field.action.edit": "Edit",
  "account.settings.editable.field.password.reset.button.support.link": "technical support",
  "account.settings.editable.field.password.reset.label": "Password",
  "account.settings.sso.link.account": "Sign in with {name}",
  "account.settings.sso.account.connected": "Linked",
  "account.settings.sso.unlink.account": "Unlink {name} account",
  "account.settings.sso.no.providers": "No accounts can be linked at this time.",
  "account.settings.sso.loading": "Loading...",
  "account.settings.sso.loading.error": "There was a problem loading linked accounts.",
  "account.settings.sso.section.header": "Linked Accounts",
  "account.settings.sso.section.subheader": "You can link your identity accounts to simplify signing in to edX.",
  "siteheader.links.courses": "Courses",
  "siteheader.links.programs": "Programs & Degrees",
  "siteheader.links.schools": "Schools & Partners",
  "siteheader.user.menu.dashboard": "Dashboard",
  "siteheader.user.menu.profile": "Profile",
  "siteheader.user.menu.account.settings": "Account",
  "siteheader.user.menu.order.history": "Order History",
  "siteheader.user.menu.logout": "Logout",
  "siteheader.user.menu.login": "Login",
  "siteheader.user.menu.register": "Sign Up",
  "app.loading.message": "Loading",
  "error.unexpected.message": "An unexpected error occurred.",
  "error.notfound.message": "The page you're looking for is unavailable or there's an error in the URL. Please check the URL and try again.",
  "app.page": "Congratulations!  You have a new micro-frontend.",
  "example.page.heading": "Example Page",
  "example.loading.message": "Loading",
  "example.loading.error": "Error: {error}",
  "feature.page.heading": "Feature Page",
  "feature.loading.message": "Loading",
  "feature.loading.error": "Error: {error}"
}