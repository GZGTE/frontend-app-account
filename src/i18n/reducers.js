import { getLocale } from './i18n-loader';
import { SET_LOCALE } from './actions';

const defaultState = {
  locale: getLocale(),
};

const reducer = (state = defaultState, action) => {
  switch (action.type) {
    case SET_LOCALE:
      return {
        locale: action.payload.locale,
      };
    default:
      return state;
  }
};

export default reducer;
