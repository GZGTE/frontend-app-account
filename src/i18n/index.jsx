import { intlShape } from 'react-intl';
import injectIntlWithShim from './injectIntlWithShim';
import {
  getCountryList,
  getCountryMessages,
  getLanguageList,
  getLanguageMessages,
  getLocale,
  getAssumedServerLanguageCode,
  getMessages,
  handleRtl,
  isRtl,
} from './i18n-loader';
import { setLocale } from './actions';
import reducer from './reducers';
import { localeSelector } from './selectors';

export {
  injectIntlWithShim as injectIntl,
  getCountryList,
  getCountryMessages,
  getLanguageList,
  getLanguageMessages,
  getLocale,
  getAssumedServerLanguageCode,
  getMessages,
  handleRtl,
  isRtl,
  intlShape,
  setLocale,
  reducer,
  localeSelector,
};
