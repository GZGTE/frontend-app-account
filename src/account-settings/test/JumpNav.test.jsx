import React from 'react';
import renderer from 'react-test-renderer';
import { IntlProvider, injectIntl } from '@edx/frontend-platform/i18n';
import { AppProvider } from '@edx/frontend-platform/react';
import { initializeMockApp, mergeConfig, setConfig } from '@edx/frontend-platform';

import { BrowserRouter as Router } from 'react-router-dom';
import JumpNav from '../JumpNav';
import configureStore from '../../data/configureStore';

const IntlJumpNav = injectIntl(JumpNav);

describe('JumpNav', () => {
  mergeConfig({
    ENABLE_DEMOGRAPHICS_COLLECTION: false,
  });

  let props = {};
  let store;

  beforeEach(() => {
    initializeMockApp({
      authenticatedUser: {
        userId: 3,
        username: 'abc123',
        administrator: true,
        roles: [],
      },
    });

    props = {
      intl: {},
      displayDemographicsLink: false,
    };
    store = configureStore({
      notificationPreferences: {
        showPreferences: false,
      },
    });
  });

  it('should not render Optional Information link', () => {
    const tree = renderer.create((
      // Had to wrap the following in a router or I will receive an error stating:
      // "Invariant failed: You should not use <NavLink> outside a <Router>"
      <Router>
        <IntlProvider locale="en">
          <AppProvider store={store}>
            <IntlJumpNav {...props} />
          </AppProvider>
        </IntlProvider>
      </Router>
    ))
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render Optional Information link', () => {
    setConfig({
      ENABLE_DEMOGRAPHICS_COLLECTION: true,
    });

    props = {
      ...props,
      displayDemographicsLink: true,
    };

    const tree = renderer.create((
      // Same as previous test
      <Router>
        <IntlProvider locale="en">
          <AppProvider store={store}>
            <IntlJumpNav {...props} />
          </AppProvider>
        </IntlProvider>
      </Router>
    ))
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
