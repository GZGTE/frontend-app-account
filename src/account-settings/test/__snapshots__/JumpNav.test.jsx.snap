// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JumpNav should not render Optional Information link 1`] = `
<div
  className="py-5 justify-content-center align-items-start text-center container-fluid"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <p
        className="text-muted"
      >
        An unexpected error occurred. Please click the button below to refresh the page.
      </p>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Try again
      </button>
    </div>
  </div>
</div>
`;

exports[`JumpNav should render Optional Information link 1`] = `
<div
  className="py-5 justify-content-center align-items-start text-center container-fluid"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <p
        className="text-muted"
      >
        An unexpected error occurred. Please click the button below to refresh the page.
      </p>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Try again
      </button>
    </div>
  </div>
</div>
`;
